File: lab6/lab6_UniqueBinarySearchTrees.py
class Solution:
    def numTrees(self, n):
        dp = [0] * (n + 1)
        
        dp[0] = 1  
        dp[1] = 1  
        
        for nodes in range(2, n + 1):
            for root in range(1, nodes + 1):
                left_trees = dp[root - 1]  
                right_trees = dp[nodes - root]  
                dp[nodes] += left_trees * right_trees
        return dp[n]

# Example usage
if __name__ == "__main__":
    solution = Solution()
    print(solution.numTrees(3))  # Output: 5
    print(solution.numTrees(1))  # Output: 1

File: labs/lab6/lab6_UniqueBinarySearchTrees.js
class Solution {
    numTrees(n) {
        const dp = new Array(n + 1).fill(0);

        dp[0] = 1; 
        dp[1] = 1; 

        for (let nodes = 2; nodes <= n; nodes++) {
            for (let root = 1; root <= nodes; root++) {
                const leftTrees = dp[root - 1]; 
                const rightTrees = dp[nodes - root]; 
                dp[nodes] += leftTrees * rightTrees;
            }
        }

        return dp[n];
    }
}

const solution = new Solution(); 
let ret = solution.numTrees(3);  
console.log(ret);  

ret = solution.numTrees(1);
console.log(ret);  
